### PR TITLE
[BEAM-2147] Re-enable UsesTimersInParDo tests on Dataflow, by fixing TestDataflowRunner and PAssertTest

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
Sad face - the build needed is not actually queued. Presumably it was killed because known breakage in `master` doomed it anyhow. Rebasing and re-triggering.

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
Noting that this PR has already been validated against Dataflow's integration suite in both batch and streaming, but we should get clear Jenkins signal on it.

**Comment by kennknowles** [comment]
R: @tgroh finally green!

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java [hunk 1]
```diff
@@ -99,88 +96,115 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
...
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      Thread.interrupted();
+      return false;
+    }
+
+    // Getting the final state may have timed out; it may not indicate a failure.
+    // This cancellation may be the second
+    if (finalState == null || !finalState.isTerminal()) {
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java]: Can `job.waitUntilFinish` return null?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/TestDataflowRunnerTest.java [hunk 2]
```diff
@@ -603,8 +584,12 @@ public void testStreamingOnSuccessMatcherWhenPipelineFails() throws Exception {
     when(mockJob.waitUntilFinish(any(Duration.class), any(JobMessagesHandler.class)))
         .thenReturn(State.FAILED);
 
-    runner.run(p, mockRunner);
-    // If the onSuccessMatcher were invoked, it would have crashed here.
+    try {
+      runner.run(p, mockRunner);
+    } catch (RuntimeException e) {
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/TestDataflowRunnerTest.java]: This looks like it should be an expected exception

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java (hunk 1 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java]: It is documented to do so on timeout or interruption, though it does seem like a design error.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/TestDataflowRunnerTest.java (hunk 2 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/TestDataflowRunnerTest.java]: It actually isn't a perfect fit - this isn't testing that an exception is thrown. Note that if an exception is _not_ thrown this test can still succeed, as long as the `onSuccessMatcher` is not invoked. But that's getting a bit silly about fine-grained testing, and the comment does "expect" the exception. So: done.

**Pull request merged**